File: backend-v5/lms/homework/Homework.php
<?php

// Modul: lms/homework
// Klasa: Homework
// Opis: jedna zadaća


require_once(Config::$backend_path."lms/homework/ProgrammingLanguage.php");

class Homework {
	public $id;
	public $name, $CourseUnit, $AcademicYear, $nrAssignments, $maxScore, $ScoringElement, $deadline, $active, $ProgrammingLanguage, $automatedTesting, $attachment, $allowedExtensions, $publishedDateTime, $text /* postavka */;
	
	public static function fromId($id) {
		$hw = DB::query_assoc("SELECT id, naziv name, predmet CourseUnit, akademska_godina AcademicYear, zadataka nrAssignments, bodova maxScore, komponenta ScoringElement, UNIX_TIMESTAMP(rok) deadline, aktivna active, programskijezik ProgrammingLanguage, automatsko_testiranje automatedTesting, attachment, dozvoljene_ekstenzije allowedExtensions, UNIX_TIMESTAMP(vrijemeobjave) publishedTime, postavka_zadace text FROM zadaca WHERE id=$id");
		if (!$hw) throw new Exception("Unknown homework $id", "404");
		
		$hw = Util::array_to_class($hw, "Homework", array("CourseUnit", "AcademicYear", "ScoringElement", "ProgrammingLanguage"));
		if ($hw->active == 1) $hw->active=true; else $hw->active=false; // FIXME use boolean in database
		if ($hw->attachment == 1) $hw->attachment=true; else $hw->attachment=false;
		if ($hw->automatedTesting == 1) $hw->automatedTesting=true; else $hw->automatedTesting=false;
		return $hw;
	}
	
	// List of homeworks that are due for submitting (deadline is near)
	// Gets no more than $limit homeworks, ordered by deadline descending
	// Doesn't return homeworks on inactive student modules
	public static function getLatestForStudent($student, $limit) {
		// FIXME u ovaj upit je ugrađena i provjera da li je aktivan studentski modul
		$results = DB::query_table("SELECT z.id id, z.naziv name, z.predmet CourseUnit, z.akademska_godina AcademicYear, z.zadataka nrAssignments, z.bodova maxScore, z.komponenta ScoringElement, UNIX_TIMESTAMP(z.rok) deadline, z.aktivna active, z.programskijezik ProgrammingLanguage, z.automatsko_testiranje automatedTesting, z.attachment, z.dozvoljene_ekstenzije allowedExtensions, UNIX_TIMESTAMP(z.vrijemeobjave) publishedTime, z.postavka_zadace text 
		FROM zadaca as z, student_predmet as sp, ponudakursa as pk, predmet as p
		WHERE z.predmet=pk.predmet AND z.akademska_godina=pk.akademska_godina AND sp.student=$student AND sp.predmet=pk.id AND pk.predmet=p.id AND z.rok>curdate() AND z.aktivna=1
		ORDER BY rok DESC
		LIMIT $limit");
		$homeworks = array();
		foreach ($results as $hw) {
			// Skip if student has passing grade (optimize?)
			$grade = DB::get("SELECT COUNT(*) FROM konacna_ocjena WHERE predmet=" . $hw['CourseUnit'] . " AND ocjena>=6");
			if ($grade > 0) continue;

			$hw = Util::array_to_class($hw, "Homework", array("CourseUnit", "AcademicYear", "ScoringElement", "ProgrammingLanguage"));
			if ($hw->active == 1) $hw->active=true; else $hw->active=false; // FIXME use boolean in database
			if ($hw->attachment == 1) $hw->attachment=true; else $hw->attachment=false;
			if ($hw->automatedTesting == 1) $hw->automatedTesting=true; else $hw->automatedTesting=false;
			
			$homeworks[] = $hw;
		}
		
		return $homeworks;
	}
	
	// List of homeworks that student submitted and were recently reviewed
	// No more than $limit homeworks, ordered by time of review descending
	public static function getReviewedForStudent($student, $limit) {
		// Not very optimal...
		$results = DB::query_table("SELECT z.id id, z.naziv name, z.predmet CourseUnit, z.akademska_godina AcademicYear, z.zadataka nrAssignments, z.bodova maxScore, z.komponenta ScoringElement, UNIX_TIMESTAMP(z.rok) deadline, z.aktivna active, z.programskijezik ProgrammingLanguage, z.automatsko_testiranje automatedTesting, z.attachment, z.dozvoljene_ekstenzije allowedExtensions, UNIX_TIMESTAMP(z.vrijemeobjave) publishedTime, z.postavka_zadace text 
		FROM zadatak as zk, zadaca as z, predmet as p 
		WHERE zk.student=$student and zk.status!=1 and zk.status!=4 and zk.zadaca=z.id and z.predmet=p.id and zk.vrijeme < SUBDATE(NOW(), INTERVAL 1 MONTH)
		ORDER BY zk.id DESC 
		LIMIT $limit");
		$repeated = array();
		$homeworks = array();
		foreach ($results as $hw) {
			if ( in_array($hw['id'], $repeated) ) continue; // ne prijavljujemo vise puta istu zadacu
			$repeated[] = $hw['id'];
			
			$hw = Util::array_to_class($hw, "Homework", array("CourseUnit", "AcademicYear", "ScoringElement", "ProgrammingLanguage"));
			if ($hw->active == 1) $hw->active=true; else $hw->active=false; // FIXME use boolean in database
			if ($hw->attachment == 1) $hw->attachment=true; else $hw->attachment=false;
			if ($hw->automatedTesting == 1) $hw->automatedTesting=true; else $hw->automatedTesting=false;
			
			$homeworks[] = $hw;
		}
		return $homeworks;
	}
	
	// List of homeworks published for course
	public static function fromCourse($courseUnitId, $academicYearId) {
		$homeworks = DB::query_table("SELECT id, naziv name, predmet CourseUnit, akademska_godina AcademicYear, zadataka nrAssignments, bodova maxScore, komponenta ScoringElement, UNIX_TIMESTAMP(rok) deadline, aktivna active, programskijezik ProgrammingLanguage, automatsko_testiranje automatedTesting, attachment, dozvoljene_ekstenzije allowedExtensions, UNIX_TIMESTAMP(vrijemeobjave) publishedTime, postavka_zadace text FROM zadaca WHERE predmet=$courseUnitId and akademska_godina=$academicYearId ORDER BY komponenta, naziv");
		foreach ($homeworks as &$hw) {
			$hw = Util::array_to_class($hw, "Homework", array("CourseUnit", "AcademicYear", "ScoringElement", "ProgrammingLanguage"));
			if ($hw->active == 1) $hw->active=true; else $hw->active=false; // FIXME use boolean in database
			if ($hw->attachment == 1) $hw->attachment=true; else $hw->attachment=false;
			if ($hw->automatedTesting == 1) $hw->automatedTesting=true; else $hw->automatedTesting=false;
		}
		return $homeworks;
	}
	
	// List of homeworks published for course
	public static function fromCourseOffering($courseOfferingId) {
		$homeworks = DB::query_table("SELECT z.id id, z.naziv name, z.predmet CourseUnit, z.akademska_godina AcademicYear, z.zadataka nrAssignments, z.bodova maxScore, z.komponenta ScoringElement, UNIX_TIMESTAMP(z.rok) deadline, z.aktivna active, z.programskijezik ProgrammingLanguage, z.automatsko_testiranje automatedTesting, z.attachment, z.dozvoljene_ekstenzije allowedExtensions, UNIX_TIMESTAMP(z.vrijemeobjave) publishedTime, z.postavka_zadace text FROM zadaca z, ponudakursa pk WHERE z.predmet=pk.predmet and z.akademska_godina=pk.akademska_godina AND pk.id=$courseOfferingId ORDER BY komponenta, naziv");
		foreach ($homeworks as &$hw) {
			$hw = Util::array_to_class($hw, "Homework", array("CourseUnit", "AcademicYear", "ScoringElement", "ProgrammingLanguage"));
			if ($hw->active == 1) $hw->active=true; else $hw->active=false; // FIXME use boolean in database
			if ($hw->attachment == 1) $hw->attachment=true; else $hw->attachment=false;
			if ($hw->automatedTesting == 1) $hw->automatedTesting=true; else $hw->automatedTesting=false;
		}
		return $homeworks;
	}
}

?>

File: rorbackend/app/models/lms/homework/homework.rb
class Lms::Homework::Homework < ActiveRecord::Base
  # Uncomment following lines if working with legacy database
  # set_table_name 'zadaca'
  # set_primary_key :id
  # alias_attribute :id, :id
  # alias_attribute :name, :naziv
  # alias_attribute :course_unit_id, :predmet
  # alias_attribute :academic_year_id, :akademska_godina
  # alias_attribute :nr_assignments, :zadataka
  # alias_attribute :score, :bodova
  # alias_attribute :deadline, :rok
  # alias_attribute :active, :aktivna
  # alias_attribute :programming_language_id, :programskijezik
  # alias_attribute :attachment, :attachment
  # alias_attribute :allowed_extensions, :dozvoljene_ekstenzije
  # alias_attribute :text, :postavka_zadace
  # alias_attribute :scoring_element_id, :komponenta
  # alias_attribute :published_date_time, :vrijemeobjave

  # Uncomment following lines if working with legacy database
  # TABLE_NAME = 'zadaca'
  # ID = TABLE_NAME + '.' + 'id'
  # NAME = TABLE_NAME + '.' + 'naziv'
  # COURSE_UNIT_ID = TABLE_NAME + '.' + 'predmet'
  # ACADEMIC_YEAR_ID = TABLE_NAME + '.' + 'akademska_godina'
  # NR_ASSIGNMENTS =  TABLE_NAME + '.' + 'zadataka'
  # SCORE =  TABLE_NAME + '.' + 'bodova'
  # DEADLINE =  TABLE_NAME + '.' + 'rok'
  # ACTIVE =  TABLE_NAME + '.' + 'aktivna'
  # PROGRAMMING_LANGUAGE_ID =  TABLE_NAME + '.' + 'programskijezik'
  # ATTACHMENT =  TABLE_NAME + '.' + 'attachment'
  # ALLOWED_EXTENSIONS =  TABLE_NAME + '.' + 'dozvoljene_ekstenzije'
  # TEXT =  TABLE_NAME + '.' + 'postavka_zadace'
  # SCORING_ELEMENT_ID =  TABLE_NAME + '.' + 'komponenta'
  # PUBLISHED_DATE_TIME =  TABLE_NAME + '.' + 'vrijemobjave'

  # Comment following lines if working with legacy database
  TABLE_NAME = 'lms_homework_homeworks'
  ID = TABLE_NAME + '.' + 'id'
  NAME = TABLE_NAME + '.' + 'name'
  COURSE_UNIT_ID = TABLE_NAME + '.' + 'course_unit_id'
  ACADEMIC_YEAR_ID = TABLE_NAME + '.' + 'academic_year_id'
  NR_ASSIGNMENTS =  TABLE_NAME + '.' + 'nr_assignments'
  SCORE =  TABLE_NAME + '.' + 'score'
  DEADLINE =  TABLE_NAME + '.' + 'deadline'
  ACTIVE =  TABLE_NAME + '.' + 'active'
  PROGRAMMING_LANGUAGE_ID =  TABLE_NAME + '.' + 'programming_language_id'
  ATTACHMENT =  TABLE_NAME + '.' + 'attachment'
  ALLOWED_EXTENSIONS =  TABLE_NAME + '.' + 'allowed_extensions'
  TEXT =  TABLE_NAME + '.' + 'text'
  SCORING_ELEMENT_ID =  TABLE_NAME + '.' + 'scoring_element_id'
  PUBLISHED_DATE_TIME =  TABLE_NAME + '.' + 'published_date_time'

  ALL_COLUMNS = [ID, NAME, COURSE_UNIT_ID, ACADEMIC_YEAR_ID, NR_ASSIGNMENTS, SCORE, DEADLINE, ACTIVE, PROGRAMMING_LANGUAGE_ID, ATTACHMENT, ALLOWED_EXTENSIONS, TEXT, SCORING_ELEMENT_ID, PUBLISHED_DATE_TIME]
  
  belongs_to :course_unit, :class_name => "Core::CourseUnit"
  belongs_to :academic_year, :class_name => "Core::AcademicYear"
  belongs_to :programming_language
  belongs_to :scoring_element, :class_name => "Core::ScoringElement"
  has_many :assignments
  
  validates_presence_of :name, :scoring_element_id
  
  before_save :set_published_date_time_to_now
  
  def self.from_id(id)
    select_columns = (Lms::Homework::Homework)::ALL_COLUMNS | [(Core::CourseUnit)::NAME]    
    homework = (Lms::Homework::Homework).where(:id => id).includes(:course_unit).select(select_columns)
    
    return homework
  end
  
  
  def self.get_latest_for_student(student_id, limit)
    select_columns = (Lms::Homework::Homework)::ALL_COLUMNS | [(Core::CourseUnit)::NAME]
    homeworks_pass = (Lms::Homework::Homework).includes(:course_unit => [{:course_offerings => :portfolios}]).where((Core::Portfolio)::STUDENT_ID => student_id).select(select_columns).order(:deadline)
    homeworks = []
    
    homeworks_pass.each do |homework|
      
      if !((Core::FinalGrade).where(:course_unit_id => homeworks[:course_unit_id], :student_id => student_id).where((Core::FinalGrade)::GRADE + '>=6').count > 0)
        
        homeworks << homework
      end
    end
    return homeworks
  end
  
  
  def self.get_reviewed_for_student(student_id, limit)
    select_columns = (Lms::Homework::Homework)::ALL_COLUMNS | [(Core::CourseUnit)::NAME]
    statuses = (Lms::Homework::Assignment)::STATUS + '!=1 AND ' + (Lms::Homework::Assignment)::STATUS + '!=4'
    homeworks_all = (Lms::Homework::Homework).joins(:assignments, :course_unit).where((Lms::Homework::Assignment)::STUDENT_ID => student_id).where(statuses).where([(Lms::Homework::Assignment)::TIME + '>?', 1.month.ago.to_s]).order((Lms::Homework::Assignment)::ID).limit(limit)
    
    homeworks_past = []
    homeworks = []
    
    homeworks_all.each do |homework|
      homeworks << homework if homeworks.include?(homework)
    end
    
    return homeworks
  end
  
  
  def self.from_course(course_unit_id, academic_year_id)
    homeworks = (Lms::Homework::Homework).where(:course_unit_id => course_unit_id, :academic_year_id => academic_year_id).order(:scoring_element_id, :name)
    
    return homeworks
  end
  
  def self.from_course_scoring_element(course_unit_id, academic_year_id)
    scoring_elements = (Core::ScoringElement).joins(:course_unit_type_scoring_elements).joins("INNER JOIN " + (Core::CourseUnitYear)::TABLE_NAME + " ON " + (Core::CourseUnitYear)::COURSE_UNIT_TYPE_ID + '=' + (Core::CourseUnitTypeScoringElement)::COURSE_UNIT_TYPE_ID).where((Core::CourseUnitYear)::ACADEMIC_YEAR_ID => academic_year_id, (Core::CourseUnitYear)::COURSE_UNIT_ID => course_unit_id, (Core::ScoringElement)::SCORING_ID => 4).order(:id)
    
    return scoring_elements
  end
  
  
  def self.update_score_for_student(student_id, course_unit_id, academic_year_id)
    portfolio = (Core::Portfolio).from_course_unit(student_id, course_unit_id, academic_year_id)
    all_homeworks = (Lms::Homework::Homework).from_course(course_unit_id, academic_year_id)
    
    total_score = []
    
    all_homeworks.each do |homework|
      for i in 1..homework[:nr_assignments]
        assignment = (Lms::Homework::Assignment).from_student_homework_number(author_id, homework_id, assign_no)
        # TODO Make hash for status codes
        if (assignment[:status] == 5 and assignment != nil)
          total_score[homework[:scoring_element_id]] += assignment[:score]
        end
      end
    end
    
    total_score.each do |scoring_element, score|
      portfolio.set_score(portfolio[:id], scoring_element, score)
    end
    
    return true
  end
  
  
  def self.get_score_from_course_unit(student_id, course_unit_id, academic_year_id, homework_id)
    score = (Core::ScoringElementScore).joins(:course_offering).where((Core::ScoringElementScore)::STUDENT_ID => student_id, (Core::CourseOffering)::COURSE_UNIT_ID => course_unit_id, (Core::CourseOffering)::ACADEMIC_YEAR_ID => academic_year_id, (Core::ScoringElementScore)::SCORING_ELEMENT_ID => homework_id)
    
    return score
  end
  
private
  def set_published_date_time_to_now
    self.published_date_time = Time.now
  end
  
end
